### ISSUE TITLE
GC: Representing literals of subtypes

### ISSUE THREAD

**Issue by dcodeIO** [WebAssembly/binaryen#3131]
I have been looking into extending `Literal`s, so these can represent an `anyref` holding a concrete function for example. So far I came up with

```cpp
class Literal {
public:
  // Mutable (super-)type of the literal.
  Type type;
  // Immutable type of the literal's payload. A literal of type `anyref` holding
  // a concrete function payload has `payloadType = Type(HeapType(func), false)`
  // with `type = anyref` for example.
  const Type payloadType;
private:
  // Immutable payload of the literal according to `payloadType`.
  union {
    // 32-bit value.
    int32_t v32;
    // 64-bit value.
    int64_t v64;
    // 128-bit value.
    uint8_t v128[16];
    // string name.
    Name name;
    // exception package.
    std::unique_ptr<ExceptionPackage> exn;
  }
```

and I'd like to get your opinions on this approach. Increases literal size from 24 to 32 bytes. To give a few examples:

* Concrete function of type `anyref`
  `type = anyref`, `payloadType=(ref func)`, `name=theName`
* Null function of type `anyref`
  `type = anyref`, `payloadType=funcref`, no value (`name=nullptr`)
* Concrete exception of type `exnref`
  `type = exnref`, `payloadType=(ref exn)`, `exn=theExn`
* Null exception of type `exnref`
  `type = exnref`, `payloadType=exnref`, no value (`exn=nullptr`)

A non-nullable `payloadType` indicates a concrete payload being present, while a nullable `payloadType` indicates `null` (no payload or a null payload present if applicable).

Also found a related bug in the process:

* `BinaryenLiteral` represents the type as an `int32_t`, but the id is `uintptr_t` and the only reason that the `sizeof(BinaryenLiteral) == sizeof(Literal)` check works is due to struct padding.

If this sounds good to you, please let me know what'd be a good scope for review :)

Outlook:

With this change, adding support for `i31ref` should be straight-forward:

* Concrete `i31ref` (note: not nullable)
  `type = i31ref`, `payloadType=i31ref`, `v32=theValue`
* Null variant
  `type=(ref null i31)`, `payloadType=(ref null i31)`, no value (`v32=ub`)

Support for `externref`s of a (or `anyref`s of any) concrete value looks like it'll require a new field as part of the union, perhaps holding an abstract id representing unique external values:

* Concrete `externref`
  `type = externref`, `payloadType=(ref extern)`, `refId=theId`
* Null `externref`
  `type = externref`, `payloadType=externref`, no value (`refId=ub`)

**Comment by dcodeIO** [comment]
The obvious alternative to this approach would be to make the interpreter pass along both a type and a literal, for example

```cpp
struct TypedLiteral {
  /** Actual (sub-)type. */
  Type type;
  /** Literal payload. */
  Literal literal;
}
```

This seemed like a surprise box regarding API surface and interpreter refactoring though, so I first investigated the former.

**Comment by tlively** [comment]
Another alternative might be to make all users of the interpreter aware of the type they expect to receive based on their context, rather than having them depend on inspecting the returned `Literal` (or some new returned structure) to determine the type. This seems like it would require the fewest and least invasive changes, but it could also turn out to be impossible if some interpreter client does not have enough context to know what type to expect.

**Comment by aheejin** [comment]
I don't fully understand why we need this. I might be missing something, so please let me know.

My understanding is, every single place (in the interpreter of wherever) that expects `anyref` literal should be also accepting subtype literals, so I thought we can keep literals with their lowest possible type. I think I tried to handle subtyping in all places I discovered, but there can be places that I missed, in which case we should handle subtype too.

**Comment by dcodeIO** [comment]
> I think we can keep literals with their lowest possible type

Hmm, would that imply that let's say a `funcref` representing a concrete function should be of type `(ref func)`, i.e. non-nullable, while only `null` function literals would be actual `funcref`? I guess that could work, though I'm somewhat stuck on how to split this work given that it all depends upon each other.

**Comment by aheejin** [comment]
Hmm, didn't think about the same type with nullable and non-nullable variant. I was mostly thinking about `funcref` < `anyref`... The lowest type thing wouldn't work with that I guess, sorry.

Can we make the same reference type with nullable and non-nullable variant share the same union field, so that both nullable funcref and non-nullable funcref literals access `Name name;` for their data? But for `anyref`, given that it can take multiple types of literals (it can either be `funcref` of `exnref` or anything..), disallowing `anyref` literals still looks like the simplest option there..

**Comment by tlively** [comment]
I agree with @aheejin's intuition that `Literal` should only contain the concrete runtime type of the value.

> > I think we can keep literals with their lowest possible type
> 
> Hmm, would that imply that let's say a `funcref` representing a concrete function should be of type `(ref func)`, i.e. non-nullable, while only `null` function literals would be actual `funcref`? I guess that could work, though I'm somewhat stuck on how to split this work given that it all depends upon each other.

This does seem a little strange at first, but I actually think it's the best thing to do. As for how to split up the work, it's totally fine to e.g. add definitions and API support for types/instructions with just a TODO in the interpreter and fuzzer. No need to get everything working end-to-end all at once.

**Comment by tlively** [comment]
We have pretty good support for reference-typed Literals now, so I'll go ahead and close this.
